fgpyo/fasta/sequence_dictionary.py

```diff
@@ -0,0 +1,480 @@
+
+    @property
+    def md5(self) -> Optional[str]:
+        return self.get(Keys.MD5)
+
+    @property
+    def assembly(self) -> Optional[str]:
+        return self.get(Keys.ASSEMBLY)
+
+    @property
+    def uri(self) -> Optional[str]:
+        return self.get(Keys.URI)
+
+    @property
+    def species(self) -> Optional[str]:
+        return self.get(Keys.SPECIES)
+
+    @property
+    def description(self) -> Optional[str]:
+        return self.get(Keys.DESCRIPTION)
+
+    @property
+    def topology(self) -> Optional[Topology]:
+        value = self.get(Keys.TOPOLOGY)
+        if value is None:
+            return None
+        return Topology[value]
+
+    def same_as(self, other: "SequenceMetadata") -> bool:
+        """Returns true if the sequences share a common reference name (including aliases), have
+        the same length, and the same MD5 if both have MD5s."""
+        if self.length != other.length:
+            return False
+        elif self.name != other.name and other.name not in self.all_names:
+            return False
+        self_m5 = self.md5
+        other_m5 = other.md5
+        if self_m5 is None or other_m5 is None:
+            return True
+        else:
+            return self_m5 == other_m5
+
+    def to_sam(self) -> Dict[str, Any]:
+        """Converts the sequence metadata to a dictionary equivalent to one item in the
+        list of sequences from `pysam.AlignmentHeader#to_dict()["SQ"]`."""
+        meta_dict: Dict[str, Any] = {
+            f"{Keys.SEQUENCE_NAME}": self.name,
+            f"{Keys.SEQUENCE_LENGTH}": self.length,
+        }
+        if len(self.attributes) > 0:
+            meta_dict = {**meta_dict, **self.attributes}
+
+        return meta_dict
+
+    @staticmethod
+    def from_sam(meta: Dict[Union[Keys, str], Any], index: int) -> "SequenceMetadata":
+        """Builds a `SequenceMetadata` from a dictionary.  The keys must include the sequence
+        name (`Keys.SEQUENCE_NAME`) and length (`Keys.SEQUENCE_LENGTH`).  All other keys from
+        `Keys` will be stored in the resulting attributes.
+
+        Args:
+            meta: the python dictionary with keys from `Keys`.  This is typically the dictionary
+                  stored in the `"SQ"` level of the two-level dictionary returned by the
+                  `pysam.AlignmentHeader#to_dict()` method.
+            index: the 0-based index to use for this sequence
+        """
+        name = meta[Keys.SEQUENCE_NAME]
+        length = meta[Keys.SEQUENCE_LENGTH]
+        attributes = copy.deepcopy(meta)
+        del attributes[Keys.SEQUENCE_NAME]
+        del attributes[Keys.SEQUENCE_LENGTH]
+        return SequenceMetadata(name=name, length=length, index=index, attributes=attributes)
+
+    def __getitem__(self, key: Union[Keys, str]) -> Any:
+        if key == Keys.SEQUENCE_NAME.value:
+            return self.name
+        elif key == Keys.SEQUENCE_LENGTH.value:
+            return f"{self.length}"
```

Comment:
@tfenne I'm not in love with this, but it is consistent with the pattern that accessing via named properties returns an appropriately typed value, whereas using this get method always returns a string.